### DISCUSSION TITLE
index pages and trailing slash

### DISCUSSION THREAD

**Discussion by zivkan** [statiqdev/Discussions#107] [Q&A]
I decided on using `index.md` for my blog posts (and naming the folder the title of my blog post), so any post-specific content (images, etc) can stay in the same directory as the blog post. For example, I have URLs such as:

https://domain.test/blog/my-first-post/  (this is index.html)
https://domain.test/blog/my-first-post/image.png

However, the document believes its canonical URL does not contain the trailing slash. Therefore, the canonical URL in the HTML header, the URL in the ATOM and RSS feeds, and links from the homepage all point to the "no slash" URL, which the web server then redirects to the slash.

How do I configure statiq to generate URLs for document that retain the trailing `/` when the document is otherwise ending with `index.html`?

edit: I just noticed on Google Webmaster Tools that I have a whole bunch of excluded pages of type "Alternative page with proper canonical tag". It appears that my pages are still getting indexed, but when I see errors or excluded pages, it makes me a bit worried, so I really would like to fix this.

**Comment by daveaglick** [comment]
Interesting question! I think canonically a trailing slash indicates that directory (I.e. the `index.html` file, which is how you're using it). In general, I usually see static hosts "hide" the index page so that URLs appear to be rooted at the parent folder path. [This can actually cause some problems of its own when file names and folder names collide](https://github.com/statiqdev/statiqdev.github.io/issues/32). But it sounds like your issue is different.

> How do I configure statiq to generate URLs for document that retain the trailing / when the document is otherwise ending with `index.html`

This is probably the heart of the question. Statiq contains a setting `LinkHideIndexPages` that toggles the hiding of the entire `index.html` file (going up to the root folder _without_ a slash), but there's no built-in way to both hide the index page but also leave the trailing slash.

I see two options (or both) going forward:
- Introduce a new setting, something like `LinkPreserveTrailingSlash`.
- Document and promote implementing a custom `ILinkGenerator` which is what's responsible for generating links. Right now the `Engine` isn't coded to accept a custom link generator implementation via DI, so that's something we could change.

I don't think either one will be too complicated, assuming I'm understanding the requirement correctly. Let me see what I can do.

**Reply by daveaglick** [reply]
Tracking in https://github.com/statiqdev/Statiq.Framework/issues/218
